_posts/2024-05-05-mutter-drm-leasing.md

```diff
@@ -0,0 +1,84 @@
+
+## Initial Attempt
+When I first attempted an implementation (base off of Dallas Strouse's work) I needed to teach myself everything. Hacking on mutter requires knowledge of the mutter internals, knowledge of GLib/GObject, and knowlege of the wayland library, none of which I had before starting to work on this project. Luckily I most of the backend work was already done, and those only thing I had to implement there was rewrite a function to get the CRTC (Cathode Ray Tube Controller, (name is the relic of the past but basically the controller for a display output) from the DRM device. To learn the control flow of the Wayland protocol I set the `WAYLAND_DEBUG` environment variable before running SteamVR after switching to a compositor that supported the protocol (Sway), which provided a debug printout of all the Wayland protocol events. The DRM lease specific output is here:
+```
+[ 763564.313] wl_registry@2.global(32, "wp_drm_lease_device_v1", 1)
+[ 763564.315]  -> wl_registry@2.bind(32, "wp_drm_lease_device_v1", 1, new id [unknown]@3)
+[ 763564.320]  -> wl_display@1.sync(new id wl_callback@4)
+[ 763586.314] wl_display@1.delete_id(4)
+[ 763586.322] wp_drm_lease_device_v1@3.drm_fd(fd 4)
+[ 763586.336] wp_drm_lease_device_v1@3.connector(new id wp_drm_lease_connector_v1@4278190080)
+[ 763586.347] wp_drm_lease_connector_v1@4278190080.name("DP-2")
+[ 763586.354] wp_drm_lease_connector_v1@4278190080.description("VLV 37288 Index HMD")
+[ 763586.360] wp_drm_lease_connector_v1@4278190080.connector_id(101)
+[ 763586.367] wp_drm_lease_connector_v1@4278190080.done()
+[ 763586.373] wp_drm_lease_device_v1@3.done()
+[ 763586.379] wl_callback@4.done(461)
+[ 763586.390]  -> wp_drm_lease_device_v1@3.create_lease_request(new id wp_drm_lease_request_v1@4)
+[ 763586.398]  -> wp_drm_lease_request_v1@4.request_connector(wp_drm_lease_connector_v1@4278190080)
+[ 763586.405]  -> wp_drm_lease_request_v1@4.submit(new id wp_drm_lease_v1@5)
+[ 763588.434] wp_drm_lease_v1@5.lease_fd(fd 4)
+```
+This can be difficult to look at so I created a simplified flowchart here:
+![DRM lease protocol simplified flowchart](../assets/2024-05-05-mutter-drm-leasing/simplified-drm-leasing-flowchart.png)
+
+The last hurdle was learning to use GLib/GObject, but that turned out to be fairly easy as it has great [documentation](https://docs.gtk.org/glib/), and I ended up really enjoying using it.
+
+With some fairly minor edits I was able to hack at the Wayland protocol implementation until the SteamVR error message about not having the DRM lease protocol subsided... but I was struck by a different issue:
+![SteamVR Shared IPC Compositor Init Failed](../assets/2024-05-05-mutter-drm-leasing/steamvr-error.png)
```

Comment:
This relatively small image seems to have text wrapping around it.

![Uploading Screenshot_20240506_105301.png…]()

could you add some spaces around it (and maybe the other images too)?

here's an example:
```suggestion

![SteamVR Shared IPC Compositor Init Failed](../assets/2024-05-05-mutter-drm-leasing/steamvr-error.png)

```

Reply:
I think it should be good now